rust/azure_iot_operations_services/src/leased_lock/lease.rs

```diff
@@ -0,0 +1,326 @@
+        }
+
+        Ok(Self {
+            state_store,
+            lease_name,
+            lease_holder_name,
+            current_fencing_token: Arc::new(Mutex::new(None)),
+            auto_renewal_cancellation_token: CancellationToken::new(),
+        })
+    }
+
+    /// Gets the latest fencing token related to the most recent lease.
+    ///
+    /// Returns either None or an actual Fencing Token (`HybridLogicalClock`).
+    /// None means that either a lease has not been acquired previously with this client, or
+    /// if a lease renewal has failed (if lease auto-renewal is used). The presence of a `HybridLogicalClock`
+    /// does not mean that it is the most recent (and thus valid) Fencing Token - in case
+    /// auto-renewal has not been used and the lease has already expired.
+    ///
+    /// # Panics
+    /// Possible panic if `std::Lock::lock()` fails.
+    #[must_use]
+    pub fn get_current_lease_fencing_token(&self) -> Option<HybridLogicalClock> {
+        self.current_fencing_token
+            .lock()
+            .expect("Could not lock mutex")
+            .clone()
+    }
+
+    async fn internal_acquire(
+        &mut self,
+        lease_expiration: Duration,
+        request_timeout: Duration,
+    ) -> Result<HybridLogicalClock, Error> {
+        let a = self
+            .state_store
+            .set(
+                self.lease_name.clone(),
+                self.lease_holder_name.clone(),
+                request_timeout,
+                None,
+                SetOptions {
+                    set_condition: SetCondition::OnlyIfEqualOrDoesNotExist,
+                    expires: Some(lease_expiration),
+                },
+            )
+            .await;
+
+        let state_store_response = a?;
+
+        if state_store_response.response {
+            self.current_fencing_token
+                .lock()
+                .expect("Could not lock mutex")
+                .clone_from(&state_store_response.version);
+
+            Ok(state_store_response.version.expect(
+                "Got None for fencing token. A lease without a fencing token is of no use.",
+            ))
+        } else {
+            *self
+                .current_fencing_token
+                .lock()
+                .expect("Could not lock mutex") = None;
+
+            Err(Error(ErrorKind::KeyAlreadyLeased))
+        }
+    }
+
+    /// Attempts to acquire a lease, returning if it cannot be acquired after one attempt.
+    ///
+    /// `lease_expiration` is how long the lease will remain held in the State Store after acquired, if not released before then.
+    /// `request_timeout` is the maximum time the function will wait for receiving a response from the State Store service, it is rounded up to the nearest second.
+    ///
+    /// Returns Ok with a fencing token (`HybridLogicalClock`) if completed successfully, or `Error(KeyAlreadyLeased)` if lease is not acquired.
```

Comment:
```suggestion
    /// Returns Ok with a fencing token (`HybridLogicalClock`) if completed successfully, or `Error` if lease is not acquired.
```
since there are a few other errors that can be returned

Reply:
Done.